digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala

```diff
@@ -0,0 +1,609 @@
+
+  def getByZoomLevel( boundingRectangle: BoundingRectangle, linkGeomSource: Option[LinkGeomSource] = None) : Seq[Seq[LightLane]] = {
+    withDynTransaction {
+      val assets = dao.fetchLanes(  boundingRectangle, linkGeomSource)
+      Seq(assets)
+    }
+  }
+
+  /**
+    * Returns linear assets for Digiroad2Api /lanes GET endpoint.
+    *
+    * @param bounds
+    * @param municipalities
+    * @return
+    */
+  def getByBoundingBox(bounds: BoundingRectangle, municipalities: Set[Int] = Set()): Seq[Seq[PieceWiseLane]] = {
+    val (roadLinks, change) = roadLinkService.getRoadLinksAndChangesFromVVH(bounds, municipalities)
+    val linearAssets = getMainLanesByRoadLinks(roadLinks, change)
+
+    LanePartitioner.partition(linearAssets, roadLinks.groupBy(_.linkId).mapValues(_.head))
+  }
+
+  protected def getMainLanesByRoadLinks(roadLinks: Seq[RoadLink], changes: Seq[ChangeInfo]): Seq[PieceWiseLane] = {
+    val mappedChanges = LaneUtils.getMappedChanges(changes)
+    val removedLinkIds = LaneUtils.deletedRoadLinkIds(mappedChanges, roadLinks.map(_.linkId).toSet)
+    val existingAssets = fetchExistingLanesByRoadLinks(roadLinks, removedLinkIds)
+
+    val timing = System.currentTimeMillis
+    val (assetsOnChangedLinks, lanesWithoutChangedLinks) = existingAssets.partition(a => LaneUtils.newChangeInfoDetected(a, mappedChanges))
+
+    val initChangeSet = ChangeSet(
+      expiredLaneIds = existingAssets.filter(asset => removedLinkIds.contains(asset.linkId)) // Get only the assets marked to remove
+                                      .map(_.id)                                             // Get only the Ids
+                                      .toSet
+                                      .filterNot( _ == 0L)                                   // Remove the new assets (ID == 0 )
+                      )
+
+    val (projectedLanes, changedSet) = fillNewRoadLinksWithPreviousAssetsData(roadLinks, assetsOnChangedLinks, assetsOnChangedLinks, changes, initChangeSet)
+
+    val newLanes = projectedLanes ++ lanesWithoutChangedLinks
+
+    if (newLanes.nonEmpty) {
+      logger.info("Finnish transfer %d assets at %d ms after start".format(newLanes.length, System.currentTimeMillis - timing))
+    }
+
+    val groupedAssets = (assetsOnChangedLinks.filterNot(a => projectedLanes.exists(_.linkId == a.linkId)) ++ projectedLanes ++ lanesWithoutChangedLinks).groupBy(_.linkId)
+    val (filledTopology, changeSet) = laneFilter.fillTopology(roadLinks, groupedAssets, Some(changedSet))
+
+    publish(eventBus, changeSet, projectedLanes)
+    filledTopology.filter(lane => lane.laneAttributes.properties.find(_.publicId == "lane_code").head.values.head.value.toString.charAt(1) == '1')
+
+  }
+
+  def publish(eventBus: DigiroadEventBus, changeSet: ChangeSet, projectedLanes: Seq[PersistedLane]) {
+    eventBus.publish("lanes:updater", changeSet)
+    eventBus.publish("lanes:generator", changeSet.generatedPersistedLanes.filter(_.id == 0L))
```

Comment:
it was discussed between me and @ricardofmsimoes that the new lanes created by the road direction change can be created in the projectedLanes process, if it was altered a bit, instead of creating a new process for the direction changes. Can anyone think of a reason not to aggregate both processes or is this an ok thing to do?

Reply:
great talk :)